src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -418,11 +424,30 @@ private MerkleTree buildMerkleTreeVisitor(
       MetadataProvider metadataProvider,
       ArtifactPathResolver artifactPathResolver)
       throws IOException, ForbiddenActionInputException {
-    MerkleTree result = merkleTreeCache.getIfPresent(nodeKey);
-    if (result == null) {
-      result = uncachedBuildMerkleTreeVisitor(walker, metadataProvider, artifactPathResolver);
-      merkleTreeCache.put(nodeKey, result);
+    // Deduplicate concurrent computations for the same node. It's not possible to use
+    // MerkleTreeCache#get(key, loader) because the loading computation may cause other nodes to be
+    // recursively looked up, which is not allowed. Instead, use a future as described at
+    // https://github.com/ben-manes/caffeine/wiki/Faq#recursive-computations.
+    var freshFuture = new CompletableFuture<MerkleTree>();
+    var priorFuture = merkleTreeCache.asMap().putIfAbsent(nodeKey, freshFuture);
+    if (priorFuture != null) {
+      try {
+        return priorFuture.join();
+      } catch (CompletionException e) {
+        Throwable cause = checkNotNull(e.getCause());
+        if (cause instanceof IOException) {
+          throw (IOException) cause;
+        } else if (cause instanceof ForbiddenActionInputException) {
+          throw (ForbiddenActionInputException) cause;
+        } else {
+          checkState(cause instanceof RuntimeException);
+          throw (RuntimeException) cause;
+        }
+      }
     }
+    MerkleTree result = uncachedBuildMerkleTreeVisitor(walker, metadataProvider,
+        artifactPathResolver);
+    freshFuture.complete(result);
```

Comment:
`freshFuture` is not marked as completed for all code path which might hang Bazel forever. Please mark it failed when throwing exceptions above.

Reply:
Actually, that's not the issue! The exceptions above only occur if `priorFuture != null`, in which case `freshFuture` is never inserted into the cache, and nobody will ever wait on it. The actual issue is that we will fail to complete `freshFuture` if `uncachedBuildMerkleTreeVisitor` throws.

Reply:
Sorry, I didn't make it clear, I meant the exceptions thrown from `uncachedBuildMerkleTreeVisitor`.